File: app/src/main/java/uk/co/agfsoft/skyweather/view/adapter/AddFavouriteAdapter.java
package uk.co.agfsoft.skyweather.view.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import uk.co.agfsoft.skyweather.R;
import uk.co.agfsoft.skyweather.model.WeatherCity;
import uk.co.agfsoft.skyweather.presenter.AddFavouritePresenter;



public class AddFavouriteAdapter extends RecyclerView.Adapter<AddFavouriteAdapter.ViewHolder> {
    private List<WeatherCity> weatherCities;
    private AddFavouritePresenter addFavouritePresenter;

    public AddFavouriteAdapter(AddFavouritePresenter addFavouritePresenter) {
        this.addFavouritePresenter = addFavouritePresenter;
        weatherCities = new ArrayList<WeatherCity>(0);
    }

    @Override
    public AddFavouriteAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View itemView = inflater.inflate(R.layout.recycler_item_add_favourite, parent, false);
        return new AddFavouriteAdapter.ViewHolder(itemView, addFavouritePresenter);
    }

    @Override
    public void onBindViewHolder(AddFavouriteAdapter.ViewHolder holder, int position) {
        WeatherCity weatherCity = weatherCities.get(position);
        holder.updateContent(weatherCity);

    }

    @Override
    public int getItemCount() {
        return weatherCities.size();
    }

    public void setWeatherCities(List<WeatherCity> weatherCities) {
        this.weatherCities = weatherCities;
    }

    protected static class ViewHolder  extends RecyclerView.ViewHolder {

        @BindView(R.id.descriptionTextView)
        TextView descriptionTextView;
        @BindView(R.id.latLongTextView)
        TextView latlongTextView;

        Context context;
        private AddFavouritePresenter addFavouritePresenter;
        private WeatherCity weatherCity;

        public ViewHolder(View itemView, AddFavouritePresenter addFavouritePresenter) {
            super(itemView);
            this.addFavouritePresenter = addFavouritePresenter;
            ButterKnife.bind(this, itemView);
            context = itemView.getContext();
        }

        public void updateContent(WeatherCity weatherCity) {
            this.weatherCity = weatherCity;
            descriptionTextView.setText(context.getString(R.string.favourite_description_format, weatherCity.getName(), weatherCity.getCountry()));
            latlongTextView.setText(context.getString(R.string.lat_long_format, weatherCity.getLatitiude(), weatherCity.getLongitude()));
        }

        @OnClick(R.id.selectButton)
        public void onCitySelected() {
            addFavouritePresenter.citySelected(weatherCity);
        }
    }

}

File: app/src/main/kotlin/uk/co/agfsoft/skyweather/adapter/AddFavouriteAdapter.kt
package uk.co.agfsoft.skyweather.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import java.util.ArrayList

import butterknife.ButterKnife
import butterknife.OnClick
import kotlinx.android.synthetic.main.recycler_item_add_favourite.view.*
import uk.co.agfsoft.skyweather.R
import model.WeatherCity
import uk.co.agfsoft.skyweather.presenter.AddFavouritePresenter


class AddFavouriteAdapter(val addFavouritePresenter: AddFavouritePresenter) : RecyclerView.Adapter<AddFavouriteAdapter.ViewHolder>() {
    private var weatherCities: List<WeatherCity>? = null

    init {
        weatherCities = ArrayList(0)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AddFavouriteAdapter.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val itemView = inflater.inflate(R.layout.recycler_item_add_favourite, parent, false)
        return AddFavouriteAdapter.ViewHolder(itemView, addFavouritePresenter)
    }

    override fun onBindViewHolder(holder: AddFavouriteAdapter.ViewHolder, position: Int) {
        val weatherCity = weatherCities!![position]
        holder.updateContent(weatherCity)

    }

    override fun getItemCount(): Int {
        return weatherCities!!.size
    }

    fun setWeatherCities(weatherCities: List<WeatherCity>) {
        this.weatherCities = weatherCities
    }

    class ViewHolder(itemView: View, private val addFavouritePresenter: AddFavouritePresenter) : RecyclerView.ViewHolder(itemView) {

        internal var context: Context
        private var weatherCity: WeatherCity? = null

        init {
            ButterKnife.bind(this, itemView)
            context = itemView.context
        }

        fun updateContent(weatherCity: WeatherCity) {
            this.weatherCity = weatherCity
            itemView.descriptionTextView!!.text = context.getString(R.string.favourite_description_format, weatherCity.name, weatherCity.country)
            itemView.latLongTextView!!.text = context.getString(R.string.lat_long_format, weatherCity.latitiude, weatherCity.longitude)
        }

        @OnClick(R.id.selectButton)
        fun onCitySelected() {
            addFavouritePresenter.citySelected(weatherCity!!)
        }
    }

}
